# Visualizing Stack Overflow Tags Using KeplerMapper
## Table of Contents
 1. [Introduction](#Introduction)
 2. [Implementation](#Implementation)
 3. [Analysis](#Analysis)
 4. [Theoretical Perspective](#Theoretical-Perspective)

# Introduction

In the [last project](https://github.com/daniel-rossano/Data-Analysis-Projects/blob/main/Persistent%20Homology%20SO%20Tags/Persistent%20Homology%20of%20Stack%20Overflow%20Tags.ipynb), I looked at a subset of popular [Stack Overflow](https://stackoverflow.com/) tags, and tried to analyze them using persistent homology. Although the previous analysis suggested there were a specific number of "clusters" of tags which are often used together in the same posts, it was lacking because there was no way to determine which tags belong to which clusters, and no way to visualize the clusters.

In this project I attempt to remedy the issues from the last attempt by analyzing tags through a different approach. [KeplerMapper](https://kepler-mapper.scikit-tda.org/en/latest/) is a Python library which implements the Mapper algorithm, a topological data analysis technique that combines dimensionality reduction, clustering, and graph networking to visualize high-dimensional data in a convenient and intuitive way. The idea is that, if there is a way to represent the SO tags in high-dimensional Euclidean space, one can use [principal component analysis](https://en.wikipedia.org/wiki/Principal_component_analysis) (PCA) to reduce dimensionality of the tags, then use KeplerMapper to cluster and visualize the tags. While there are too many technical details to explain here, I will briefly explain the main ideas behind each step of the process as I go through them.

The motivation for this project is that, by understanding which SO tags are clustered together, one can recommend new, relevant tags to users if they've interracted or created posts having tags in the same cluster. Such a recommendation system can be used to increase user engagement.

Also, if you are not doing so already, it is recommended to view this project in [NBViewer](https://nbviewer.org/github/daniel-rossano/Data-Analysis-Projects/blob/main/Visualizing%20Stack%20Overflow%20Tags/Visualizing%20Stack%20Overflow%20Tags%20Using%20KeplerMapper.ipynb), so that one can easily see and interact with the output. That is the only cell that needs to be run if one wishes to see the visualization of tags.

# Implementation

The implementation of this project is surprisingly simple; it is the meaning behind the code that requires more explanation. For our first step, I want to represent the SO tags in Euclidean space. In theory, I could simply represent the tags in two or three dimensions and skip the PCA, but it is difficult to find two or three distinct yet meaningful ways to quantify the SO tags. Instead, I can represent the tags in high-dimensional space by describing tags in terms of their usages and co-occurrences with other tags. For example, if there were only three tags to analyze, I would represent each of them in $\mathbb{R}^3$ by:

$$\text{tag}_1 = (\text{usages of tag}_1, \text{co-occurrences of tag}_1 \text{ and tag}_2, \text{co-occurrences of tag}_1 \text{ and tag}_3)$$
$$\text{tag}_2 = (\text{co-occurrences of tag}_2 \text{ and tag}_1, \text{usages of tag}_2, \text{co-occurrences of tag}_2 \text{ and tag}_3)$$ 
$$\text{tag}_3 = (\text{co-occurrences of tag}_3 \text{ and tag}_1, \text{co-occurrences of tag}_3 \text{ and tag}_2, \text{usages of tag}_3).$$ 

This project must therefore be restricted to a random sample of $50,000$ posts, because SO does not archive co-occurrences of tags across the whole site, and I need to calculate them manually. The project is restricted specifically to $50,000$ posts because that is the limit to how many outputs one can get using the [Stack Exchange Data Explorer](https://data.stackexchange.com/stackoverflow/query/new) to obtain this sample. This contrasts with [StackAPI](https://stackapi.readthedocs.io/en/latest/) which, despite its convenience, limits queries to $10,000$ per day. 

To make the best use of this analysis, I will filter out a lot of unpopular tags so that they do not cloud up the visualization or waste computation time. Just as I did last time, I will start with a sample of $50,000$ random posts to study tags from, and then only consider tags that have been used in *at least* $100$ posts, i.e., $0.02\%$ of the posts in the sample. As I've learned new things in Python, I've also found a more efficient way to do this than last time. For starters, it is not necessary look at the $50,000$ most popular tags and cross-reference with their usages over the posts. By querying the Data Explorer to report back only the tags used in the random selection of $50,000$ posts, I can just read in the sample using pandas and split the string of tags accordingly.

In [14]:
import pandas as pd
import re
posts = pd.read_csv("SampleofPosts.csv")

all_tags = {} #used to store usages of the tags so it will be easier to sieve out tags used in less than 100 posts
relevant_tags = {} #used to store the usages of only the relevant tags, i.e., only those used in >= 100 posts
count_of_pairs = {} #used to store all co-occurrences of relevant tags


#iterate through each row, and use re.findall with the given pattern to split a given row into a list of tags
for tags in posts["Tags"]:
    tag_list = re.findall(r'<(.*?)>', tags)
    
    #go through each tag in the tag list
    for tag in tag_list:
        if tag in all_tags: #if already accounted for the tag in question, increment its usage count
            all_tags[tag] += 1
        else:
            all_tags[tag] = 1 #else, this tag is being stored for the first time



#search through all_tags, and store only the tags used >=100 times in the sample
for tag, count in all_tags.items():
    if count >= 100:
        relevant_tags[tag] = count


#and now I might as well set all_tags to None since I don't need it anymore
all_tags = None

In [15]:
from itertools import combinations #helps us efficiently iterate through combinations of tags

tag_pairs = {} #used to store pairs of tags as keys and their co-occurrences as values

#have to check co-occurrences by reading in tags for each post, just as before
for tags in posts["Tags"]:
    tag_list = re.findall(r'<(.*?)>', tags)
    
    #for all unique pairs of tags in the tag list, if tag1 and tag2 are relevant tags...
    for tag1, tag2 in combinations(tag_list, 2):
        if tag1 in relevant_tags and tag2 in relevant_tags:
            
            #...check if tag1 is lexicographically larger than tag2 to weed out redundant computations
            if tag1 > tag2: 
                tag1, tag2 = tag2, tag1
                
            #and create a tuple of the two tags to act as a key, then proceed as before
            pairing = (tag1, tag2) 
            if pairing in tag_pairs:
                tag_pairs[pairing] += 1
            else:
                tag_pairs[pairing] = 1

Now that I've found and stored all the tag usages and co-occurrences, I can start analyzing the tags. A quick check: 

In [16]:
print(len(relevant_tags.keys()))

187


shows that there are $187$ tags to work with, and therefore each tag will be represented as a tuple in $\mathbb{R}^{187}$. Visualizing the tags right now is, of course, not possible for anyone living in less than $187$ dimensions. So naturally the next step is to reduce dimension. For this project, I decided on using PCA because
1. it maximizes variance and is based on linear transformations, so it will be better at preserving structure than say, t-SNE or UMAP
2. it is easy to implement thanks to libraries like sklearn

The basic idea is that I'll run sklearn's `PCA`  on the tag-tuples, reducing it to a three-dimensional space by keeping only the three most important components in each tuple, and then project into this three-dimensional space using KeplerMapper.

To use `PCA`, I will use `relevant_tags` and `tag_pairs` to create a 2D numpy array, where each row is represented by one of the tag-tuples. This can be thought of as a $187 \times 187$ matrix, where entry $i,j$ is the usage/co-occurrence of $\text{tag}_i$ with $\text{tag}_j$.

In [17]:
import numpy as np
from sklearn.decomposition import PCA

#use a variable to list all the relevant tags; will need this later to label members of clusters, anyway
tag_list = list(relevant_tags.keys())
#initialize the 2D array as n x n
tag_matrix = np.zeros((len(tag_list), len(tag_list)))

#this will be less painful if indices of the matrix are tracked
#so create a simple mapping of tags to sorted indices in the tag list
index_map = {tag: index for index, tag in enumerate(tag_list)}

#now easily fill in the diagonals
for tag, count in relevant_tags.items():
    i = index_map[tag] #find the index of the tag to put the count in the right entry
    tag_matrix[i,i] = count

#fill in the non-diagonal entries analogously, keeping in mind this is a symmetric matrix
for (tag_A, tag_B), count in tag_pairs.items():
    i = index_map[tag_A]
    j = index_map[tag_B]
    tag_matrix[i,j] = count
    tag_matrix[j,i] = count

#no need for index_map anymore
index_map = None

#initialize sklearn's PCA, remembering to keep 3 principal components
pca = PCA(n_components = 3)
#now fit_transform the tag matrix and store the resulting array
tag_pca = pca.fit_transform(tag_matrix)

One last thing to consider is the choice of clustering algorithm. The Mapper algorithm works by projecting a high-dimensional data set into a low-dimensional data set, choosing a cover for the low-dimensional data set (KM only allows for cubes at this point in time), and then clustering points locally based on their location in the cover. Next the Mapper algorithm constructs a graph where each node represents a cluster, and whenever a point lies inside more than one cluster (which may happen due to overlap among cubes), an edge will be drawn between the corresponding nodes. For the record, KM allows for *any* projection function, and *any* clustering algorithm or distance metric. I chose to use k-means clustering for this as that seems to be the standard for text clustering. Fortunately, sklearn also provides an algorithm for k-means clustering. 

I will initialize the graph formed by the Mapper algorithm in one line. To make it is easy to track which tags belong to which cluster(s), I will form a numpy array of all the tags and use this array in the `custom_tooltips` parameter, so that KM labels the tags in the correct order. I will also separate the tags by square brackets so they are easier to read on the graph. Finally, I will use `KeplerMapper.visualize` to output a .html file of the graph and begin the analysis. Because this project is written in Jupyter Notebook, I must take the extra step of importing KM's `jupyter` to create a custom display of the .html file using the notebook.

If you are interested in exploring the KM output yourself, simply download this notebook (or use Binder) and run, being sure to uncomment the first `IFrame` line. If you are viewing this project on GitHub, it is recommended [you use NBViewer](https://nbviewer.org/github/daniel-rossano/Data-Analysis-Projects/blob/main/Visualizing%20Stack%20Overflow%20Tags/Visualizing%20Stack%20Overflow%20Tags%20Using%20KeplerMapper.ipynb) so you can easily run the notebook and view the .html.

In [18]:
import sklearn
import kmapper as km


#initialize mapper
mapper = km.KeplerMapper()
#and now build the graph; cluster using specified # of kmeans clusters; build cover w/ % overlap
graph = mapper.map(tag_pca, tag_matrix, clusterer = sklearn.cluster.KMeans(n_clusters = 6), 
                   cover = km.Cover(n_cubes = 20, perc_overlap = 0.3))

#to construct the array just modify the existing tag list and convert to np array
#the conditional is here to prevent adding multiple sets of brackets upon re-running this cell
if tag_list[0].find('[') == -1:
    tag_list = ['[' + tag + ']' for tag in tag_list]
labels = np.array(tag_list)
#finally, visualize and output the graph
mapper.visualize(graph, path_html="Visualization-of-SO-Tags.html", title = 'Clusters of Stack Overflow Tags',
                custom_tooltips = labels)

#for local use:
from IPython.display import IFrame
IFrame(src = "Visualization-of-SO-Tags.html", width=800, height=600)

In [19]:
#run this cell to view output; no need to run any other cells unless you are curious and have data experiment with!
from IPython.display import IFrame #used to display the .html in the notebook environment
IFrame(src ="https://nbviewer.org/github/daniel-rossano/Data-Analysis-Projects/blob/main/Visualizing%20Stack%20Overflow%20Tags/Sample%20and%20Visualization/Visualization-of-SO-Tags.html", width=800, height=600)

# Analysis

With the output displayed, all that remains is to ask, "what does it tell us about the Stack Overflow tags?" For starters, one should keep in mind that there are *many* different ways to change the simplicial complex generated by KM: projection function, filtering function and its number of clusters or definition of closeness, number of cubes, and percentage of cube overlap can all dramatically change the output with even the slightest adjustment. The choices of projection and filtering function were already justified, though they may not be perfect! As for why I used $6$ k-means clusters and $20$ cubes with $30\%$ overlap, it was essentially the result of trial-and-error, with these parameters yielding the most interesting results.

In short, I believe that the current output does suggest a few relationships among SO tags, however there are likely other methods of analyzing these tags. Unfortunately there are some "mega nodes" which contain such a vast majority of tags that they cannot possibly be described by one concept, programming language, or computer science discipline. Attempts to "break up" these clusters into smaller, more descriptive clusters by changing the parameters mentioned above did not seem any more productive. Additionally, there are tags that one would expect to be incredibly popular, like "C++," that appear to be isolated from every other tag, which suggests not all important features of the data were captured in this process. If one wishes to improve upon the analysis here, it should be noted that it is possible  PCA is not the optimal method of dimensionality reduction for this data set. Or perhaps another, non-linear clustering algorithm like DBSCAN or t-SNE may be more useful for further analysis.

On a more positive note, in between the "isolated nodes" and "mega nodes," there are some smaller nodes having $2-25$ members with commonalities. Of particular interest may be the "web-dev node" of $25$ members, containing tags such as "xml," "sqlite," and "android-studio," as the majority of its entries refer directly to technologies, environments, and protocols used in web development. This finding seems to suggest that the work done here is not in vain, but rather in need of further exploration.


# Theoretical Perspective
This section is best suited for the mathematically inclined, and involves only theoretical ideas. The analysis above serves as the true summary of this project. But one may wonder how well the analysis of tags using KeplerMapper holds up against the [analysis using persistent homology](https://github.com/daniel-rossano/Data-Analysis-Projects/blob/main/Persistent%20Homology%20SO%20Tags/Persistent%20Homology%20of%20Stack%20Overflow%20Tags.ipynb) (PH). Although the overall number of clusters differs quite a bit from the PH approach, this is mostly due to the choice of deciding which clusters counted as sufficiently persistent. In theory, both approaches are actually quite similar; understanding why requires some mathematical understanding of what is happening in each project.

For the PH approach, I clustered tags by defining an abstract finite metric space $M = (X,d)$, where $X$ is a set whose elements are the set $P$ of all posts in the sample which are tagged with some tag $T$, and $d$ is the Jaccard metric, defined by $d(P,Q) = 1 - \frac{|P \cap Q|}{|P \cup Q|}$ (here $| \cdot |$ denotes the cardinality of a set).

In the analysis using KeplerMapper, I embedded the tags into $\mathbb{R}^n$ for $n = |X|$, using a finite set $Y$, whose elements are tuples are of the form $y_i = (|X_i \cap X_1|, |X_i \cap X_2|, \ldots , |X_i \cap X_n|)$, and where $X_i$ is a fixed element of $X$ and $X_1,X_2,\ldots,X_n$ are the $n$ elements of $X$. Any clustering algorithm will apply the Euclidean metric $| \cdot |$ on $Y$, and therefore the difference between the two projects is that I've gone from working with the abstract metric space $M$ to working with the Euclidean metric space $N = (Y, | \cdot |)$. In particular, this means the results will be similar if there is some way to preserve the notion of "closeness" between elements of $M$ and elements of $N$. One can formalize the setup for this problem mathematically by the following:

Let $X = \{X_i\}^{i=n} _{i=1}$ be a finite collection of nonempty finite sets, and let $M = (X, d)$ be the finite metric space formed by equipping $X$ with the metric $d(X_i, X_j) = 1 - \frac{|X_i \cap X_j|}{|X_i \cup X_j|}$. Now let $Y \subset \mathbb{R}^n$ be a finite set consisting of all tuples of the form $y_i = (|X_i \cap X_1|, |X_i \cap X_2|, \ldots , |X_i \cap X_n|)$ for each $1 \leq i \leq n$. 

Ideally, I would like to find a continuous function $f:M \to N$ that can preserve certain properties of the clusters. Truly expanding on this idea and proving it are certainly items of interest, but this will need to be done at some point in the future when I have more time to do so.